share/oaipmh/views.py

```diff
@@ -0,0 +1,256 @@
+            'identifier_scheme': 'oai',
+            'repository_identifier': self.REPOSITORY_IDENTIFIER,
+            'identifier_delimiter': self.IDENTIFER_DELIMITER,
+            'sample_identifier': self.oai_identifier(1),
+        })
+
+    def _do_listmetadataformats(self, kwargs):
+        self.context['formats'] = self.FORMATS
+        if 'identifier' in kwargs:
+            self.resolve_oai_identifier(kwargs['identifier'])
+
+    def _do_listsets(self, kwargs):
+        if 'resumptionToken' in kwargs:
+            self.errors.append(oai_errors.BadResumptionToken(kwargs['resumptionToken']))
+            return
+        self.context['sets'] = Source.objects.values_list('name', 'long_title')
+
+    def _do_listidentifiers(self, kwargs):
+        works = self._load_page(kwargs)
+        self.context['records'] = [self._header_context(work) for work in works]
+
+    def _do_listrecords(self, kwargs):
+        works = self._load_page(kwargs)
+        self.context['records'] = [(
+            self._header_context(work),
+            self.context['format'].work_context(work, self)
+        ) for work in works]
+
+    def _do_getrecord(self, kwargs):
+        format = self._get_format(kwargs['metadataPrefix'])
+        work = self.resolve_oai_identifier(kwargs['identifier'])
+        if self.errors:
+            return
+        self.context.update({
+            'header': self._header_context(work),
+            'format': format,
+            'work': format.work_context(work, self),
+        })
+
+    def _get_format(self, prefix):
+        try:
+            return next(f for f in self.FORMATS if f.prefix == prefix)()
+        except StopIteration:
+            self.errors.append(oai_errors.BadFormat(prefix))
+
+    def _header_context(self, work):
+        return {
+            'oai_identifier': self.oai_identifier(work),
+            'datestamp': format_datetime(work.date_modified),
+            'set_specs': work.sources.values_list('source__name', flat=True),
+        }
+
+    def _load_page(self, kwargs):
+        if 'resumptionToken' in kwargs:
+            try:
+                queryset, next_token, prefix, cursor = self._resume(kwargs['resumptionToken'])
+            except ValueError:
+                self.errors.append(oai_errors.BadResumptionToken(kwargs['resumptionToken']))
+                return []
+        else:
+            queryset = self._record_queryset(kwargs)
+            next_token = self._get_resumption_token(kwargs)
+            prefix = kwargs['metadataPrefix']
+            cursor = 0
+        self.context['format'] = self._get_format(prefix)
+        if self.errors:
+            return []
+        if not queryset.exists():
+            self.errors.append(oai_errors.NoResults())
+            return []
+        # TODO is there a way to prefetch Sources/Relations/Identifiers just for this slice? https://code.djangoproject.com/ticket/26780
+        works = queryset[cursor:cursor + self.PAGE_SIZE + 1]
```

Comment:
There is! We just added it into the OSF to fix performance issues. I think I'm going to release it as a pypi package, it seems fairly useful. I'll keep you posted?